### PR TITLE
Fixed a bug where a newline was treated as a valid value even if it was included at the end.

### PR DESCRIPTION
**Pull request by ytetsuro**
**Description**
I want it to be false when

```php
<?php
$validation->check("0\n", 'decimal');
```

However, the regular expression validation implemented in CodeIgniter4 allowed a trailing newline code.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by MGatner** [comment]
It took me a minute but I think I understand this now. Basically, because we used `^` and `$` to check "beginning/end of *line*" the newlines were not included in the validation. Switching to `\A` and `\z` (https://www.php.net/manual/en/regexp.reference.anchors.php) matches based on the entire *string* which will handle newline as a character to be validated as well.

Would implementing the `PCRE_MULTILINE` modifier (https://www.php.net/manual/en/reference.pcre.pattern.modifiers.php) be a better solution? I'm not great with regex.

**Comment by MGatner** [comment]
@ytetsuro Please squash this to a single commit. Alternatively, whoever merges it please be sure to squash.

**Comment by ytetsuro** [comment]
> Please squash this to a single commit.

fixed.

**Comment by ytetsuro** [comment]
> better solution?

It is I think a bad solution.
If you enable PCRE_MULTILINE modifier, it will be enabled when following.

```php
<?php
$foo = <<<EOT
0123
234
567
EOT;
preg_match('/^[0-9]+$/m', $foo); // true
```

**Comment by MGatner** [comment]
Gotcha, thanks for clarifying! I'm good with this solution then, we'll see if anyone else weighs in.

**Review by MGatner — APPROVED** [review]

**Review by paulbalandan — APPROVED** [review]

File: system/Validation/FormatRules.php
Comment (samsonasik) [inline system/Validation/FormatRules.php:47]: please add `@see https://regex101.com/` link comment so it will be easier to verify in the future if bug/improvement needed

**Review by michalsn — APPROVED** [review]

File: system/Validation/FormatRules.php
Comment (ytetsuro) [reply system/Validation/FormatRules.php:47]: @samsonasik 

> please add @see https://regex101.com/ link comment so it will be easier to verify in the future if bug/improvement needed

Do we write this link in the code where all the regular expressions are written?
I don't see the need to add a link.
I think the attached URL is maintained by an individual, but how long will it be valid?

Comment (samsonasik) [reply system/Validation/FormatRules.php:47]: You can check these:

https://github.com/codeigniter4/CodeIgniter4/blob/3119fc8d0b8790cc408c292bcab218931f6d32d2/system/View/Parser.php#L576

https://github.com/codeigniter4/CodeIgniter4/blob/3119fc8d0b8790cc408c292bcab218931f6d32d2/system/View/Parser.php#L742

https://github.com/codeigniter4/CodeIgniter4/blob/3119fc8d0b8790cc408c292bcab218931f6d32d2/utils/Rector/UnderscoreToCamelCaseVariableNameRector.php#L28

Yes, not everwhere yet, but I think it will be better to add it whenever possible for better future.

I've no information about how long it will be valid, but it seems it long enough ,for example, this is from 5 years ago https://regex101.com/library/wL5mP7 .

Comment (ytetsuro) [reply system/Validation/FormatRules.php:47]: Will this link really be easy to maintain?
Every time a regular expression is edited, does the URL also need to be maintained, and will all the maintainers review and manage it?

The system also has the ability to delete the URLs that are created.
After the merge, if the link is deleted, the information will be lost.
I think test code is more constructive than a link that may or may not work.

What do other MEMBERS think about this?
I don't want to add this URL as my commit, but I will reluctantly add it if many members agree with this.

Comment (sfadschm) [reply system/Validation/FormatRules.php:47]: I vote for adding the link.
As it is positioned directly above the code I would think it is implicit to maintain it (or at least see what it does) when changing the regex itself.
Also, while of course the link might point to a dead end some future day, I consider it to be better to provide information that might get lost instead of not providing info at all in the first place.

Comment (michalsn) [reply system/Validation/FormatRules.php:47]: It would be nice to have a link. It will make our life easier in the future.

Comment (ytetsuro) [reply system/Validation/FormatRules.php:48]: Thanks for giving me your input.
I have added your comment.

**Pull request merged**
